src/v/cloud_storage/remote_partition.cc

```diff
@@ -523,66 +524,71 @@ class partition_record_batch_reader_impl final
-                      spillovers.empty()
-                      || spillovers.get_max_timestamp_column()
-                             .last_value()
-                             .value_or(model::timestamp::max()())
-                           >= query_ts()) {
-                        vlog(
-                          _ctxlog.debug,
-                          "Manifest query raced with retention and the result "
-                          "is below the clean/start offset for {}",
-                          query_ts);
-                        return true;
-                    }
-
-                    // query was not meant for archive region. fallthrough and
-                    // log an error
-                    return false;
-                })) {
-                // error was handled
-                co_return;
+            // Out of range queries are unexpected. The caller must take care
+            // to send only valid queries to remote_partition. I.e. the fetch
+            // handler does such validation. Similar validation is done inside
+            // remote partition.
+            //
+            // Out of range at this point is due to a race condition or due to
+            // a bug. In both cases the only valid action is to throw an
+            // exception and let the caller deal with it. If the caller doesn't
+            // handle it it leads to a closed kafka connection which the
+            // end clients retry.
+            if (cur.error() == error_outcome::out_of_range) {
+                ss::visit(
+                  query,
+                  [&](model::offset) {
+                      vassert(
+                        false,
+                        "Unreachable code. Remote partition doesn't know how "
+                        "to "
+                        "handle model::offset queries.");
+                  },
+                  [&](kafka::offset query_offset) {
+                      // Bug or retention racing with the query.
+                      const auto log_start_offset
+                        = _partition->_manifest_view->stm_manifest()
+                            .full_log_start_kafka_offset();
+
+                      if (
+                        log_start_offset && query_offset < *log_start_offset) {
+                          vlog(
+                            _ctxlog.warn,
+                            "Manifest query below the log's start Kafka "
+                            "offset: "
+                            "{} < {}",
+                            query_offset(),
+                            log_start_offset.value()());
+                      }
+                  },
+                  [&](model::timestamp query_ts) {
+                      // Special case, it can happen when a timequery falls
+                      // below the clean offset. Caused when the query races
+                      // with retention/gc.
+                      auto const& spillovers = _partition->_manifest_view
```

Comment:
I don't think we should. It keeps the implementation simpler. We can't guarantee that even if we didn't detect a truncation at this point there won't be one before we bubble this response to kafka layer, or network, or client, or before client uses the offset.

Let's keep it simple unless we can show that this happens often and it would improve performance (usability a teeny bit too?) if we do internal retries (avoiding the overhead of server-client-server roundtrip in an e2e retry).

This is akin to the "End-to-end principle"
> The end-to-end principle is a design framework in [computer networking](https://en.wikipedia.org/wiki/Computer_network). In [networks designed](https://en.wikipedia.org/wiki/Network_planning_and_design) according to this principle, guaranteeing certain application-specific features, such as reliability and security, requires that they reside in the communicating [end nodes](https://en.wikipedia.org/wiki/End_node) of the network. Intermediary nodes, such as [gateways](https://en.wikipedia.org/wiki/Gateway_(telecommunications)) and [routers](https://en.wikipedia.org/wiki/Router_(computing)), that exist to establish the network, may implement these to improve efficiency but cannot guarantee end-to-end correctness.
> —https://en.wikipedia.org/wiki/End-to-end_principle